src/vs/workbench/contrib/chat/browser/chatInputPart.ts

```diff
@@ -595,6 +603,64 @@ export class ChatInputPart extends Disposable implements IHistoryNavigationWidge
 				});
 				widget.ariaLabel = ariaLabel;
 				widget.tabIndex = 0;
+			} else if (attachment.isImage) {
+				let buffer = new Uint8Array(0);
+				if (attachment.value instanceof URI) {
+					const readFile = await this.fileService.readFile(attachment.value);
+					buffer = readFile.value.buffer;
+				} else {
+					buffer = attachment.value as Uint8Array;
+				}
+
+				const blob = new Blob([buffer], { type: 'image/png' });
+				const url = URL.createObjectURL(blob);
+
+				const img = document.createElement('img');
+				img.classList.add('chat-attached-context-image');
+				img.src = url;
+				img.alt = '';
+
+				// Pill with tiny image
+				const pill = document.createElement('div');
+				pill.classList.add('chat-attached-context-pill');
+
+				const pillImg = document.createElement('img');
+				pillImg.src = url;
+				pillImg.alt = '';
+				pillImg.classList.add('chat-attached-context-pill-image');
+
+				pill.appendChild(pillImg);
+
+				// Custom label
+				const textLabel = document.createElement('span');
+				textLabel.textContent = attachment.name;
+				textLabel.classList.add('chat-attached-context-custom-text');
+
+				widget.style.position = 'relative';
+				widget.appendChild(pill);
+				widget.appendChild(textLabel);
+
+				const ariaLabel = localize('chat.imageAttachment', "Attached image, {0}", attachment.name);
+
+				widget.ariaLabel = ariaLabel;
+				widget.tabIndex = 0;
+
+				// Hover
+				const hoverElement = document.createElement('div');
+				hoverElement.appendChild(img);
+				hoverElement.classList.add('chat-attached-context-hover');
+				hoverElement.setAttribute('aria-label', ariaLabel);
+
+				this._register(this.hoverService.setupManagedHover(hoverDelegate, widget, hoverElement));
```

Comment:
You can pass a callback so the content doesn't have to be computed until the hover happens. Can you do it that way and avoid reading the file immediately? As is, it will read all the files from disk for every attachment, every time an attachment is added or removed, and that's unnecessary work. I also don't like this loop being async. There's nothing preventing `initAttachedContext` from running twice and having two async loops interfering with each other.

Reply:
This shouldn't be `this._register`, it should go on this.attachedContextDisposables